### PR TITLE
[feature-vpt-integration-1675] Add support for applying ARC override when DMARC fails

### PR DESCRIPTION
**Pull request by markeissler**
This PR is part of the following series of sequentially dependent (stacked) branches. Please review and comment on changes in each PR individually.

**NOTE 1: BRANCHES MUST be merged in reverse order (1677-->1521) to avoid conflicts.**

**NOTE 2: PR number order and merge order are not necessarily the same. Verify the name of the PR before merging into its parent branch.**

The stack includes (sequentially, in reverse):

- feature-vpt-integration-1677
- feature-vpt-integration-1675
- feature-vpt-integration-1674
- feature-vpt-integration-1521-1
- feature-vme-integration-1521

This specific PR adds support for applying an override, when DMARC would otherwise fail, if validation of an ARC chain passes:

- Add support for evaluating ARC status when DMARC fails
- Enforce ARC override behavior to require whitelist pass
- Enforce that only one trusted A-R header with `arc=pass` exists, otherwise set `arc=fail`
- Update docs and copyright years for touched files

### CODE REVIEW

File: opendmarc/Makefile.am [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2012, The Trusted Domain Project.  All rights reserved.
+# Copyright (c) 2012, 2018, The Trusted Domain Project.  All rights reserved.
```
Comment (mskucherawy) [inline opendmarc/Makefile.am]: But nothing changed...

File: opendmarc/opendmarc.c [hunk 2]
```diff
@@ -775,18 +778,18 @@ dmarcf_freearray(char **a)
 }
 
 /*
-**  DMARCF_MKARRAY -- convert a comma-separated string into an array
+**  DMARCF_MKARRAY -- convert a delimiter-separated string into an array
 **
 **  Parameters:
 **  	str -- input string
-**  	array -- output array
+**  	delim -- delimeter string
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Given the definition of strtok_r(), this should actually be "set of delimiter characters".

File: opendmarc/opendmarc.c [hunk 3]
```diff
@@ -2509,8 +2520,54 @@ mlfi_eom(SMFICTX *ctx)
...
 			{
-				if (ar.ares_result[c].result_result == ARES_RESULT_PASS)
-					dfc->mctx_arcpass = TRUE;
+				limit_arc++;
+				/*
+				** If we already came across a trusted A-R header with arc=pass
+				** we need to fail.
+				*/
+				if (ar.ares_result[c].result_result == ARES_RESULT_PASS && limit_arc == 1)
+				{
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Braces not needed when both blocks are single-line.

File: opendmarc/opendmarc.c [hunk 4]
```diff
@@ -2509,8 +2520,54 @@ mlfi_eom(SMFICTX *ctx)
...
+					dfc->mctx_arcpass = ARES_RESULT_FAIL;
+				}
+
+				/*
+				** Check arc status against whitelist policy
+				*/
+				if (dfc->mctx_arcpass == ARES_RESULT_PASS && g_hash_table_size(domain_whitelist_hash) > 0)
+				{
+					u_char *arcchain = NULL, *arcdomain;
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Each on its own line, please.

File: opendmarc/opendmarc.c [hunk 5]
```diff
@@ -2509,8 +2520,54 @@ mlfi_eom(SMFICTX *ctx)
...
+					int arcchainlen = 0, arcchainitempass = 0;
+
+					for (pc = 0;
+						pc < ar.ares_result[c].result_props;
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Fix indent.

File: opendmarc/opendmarc.c [hunk 6]
```diff
@@ -2509,8 +2520,54 @@ mlfi_eom(SMFICTX *ctx)
...
+						pc++)
+					{
+						if (ar.ares_result[c].result_ptype[pc] == ARES_PTYPE_ARCCHAIN)
+						{
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Braces not needed.

File: opendmarc/opendmarc.c [hunk 7]
```diff
@@ -2509,8 +2520,54 @@ mlfi_eom(SMFICTX *ctx)
...
+							arcchain = ar.ares_result[c].result_value[pc];
+						}
+					}
+					if (arcchain != NULL)
+					{
+						arcchainlen = dmarcf_mkarray(arcchain, ":",
+						                                     &dfc->mctx_arcchain);
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Align indent.

File: opendmarc/opendmarc.c [hunk 8]
```diff
@@ -2509,8 +2520,54 @@ mlfi_eom(SMFICTX *ctx)
...
+						for (pc = 0; dfc->mctx_arcchain[pc] != NULL; pc++)
+						{
+							arcdomain = (u_char *)g_utf8_strdown(dfc->mctx_arcchain[pc],
+							                                     strlen(dfc->mctx_arcchain[pc]));
+							g_hash_table_contains(domain_whitelist_hash,
+							                      arcdomain) && arcchainitempass++;
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: WAT

File: opendmarc/opendmarc.c [hunk 9]
```diff
@@ -2509,8 +2520,54 @@ mlfi_eom(SMFICTX *ctx)
...
+						}
+						if (arcchainlen == arcchainitempass)
+						{
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Braces not needed.

File: opendmarc/opendmarc.c [hunk 10]
```diff
@@ -3073,20 +3130,14 @@ mlfi_eom(SMFICTX *ctx)
 	  case DMARC_POLICY_ABSENT:		/* No DMARC record found */
 	  case DMARC_FROM_DOMAIN_ABSENT:	/* No From: domain */
 		aresult = "none";
-		// ret = SMFIS_ACCEPT;
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Please just remove the lines.

File: opendmarc/opendmarc.c [hunk 11]
```diff
@@ -3162,15 +3213,33 @@ mlfi_eom(SMFICTX *ctx)
...
+	** information is valid: arc=pass, arc.chain is present, and all listed
+	** domains in the chain are whitelisted.
+	**
+	** Additional logging is provided when DMARC is in failure mode and arc=pass
+	** but authentication still fails because of an invalid arc.chain to assist
+	** with administrative debugging.
+	*/
+	if (result == DMARC_RESULT_REJECT
+	    && dfc->mctx_arcpass == ARES_RESULT_PASS
+		&& dfc->mctx_arcpolicypass != DMARC_ARC_POLICY_RESULT_PASS
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Alignment, plus the continued operator should be at the end of the line:
```
if (result == THING &&
    condition2 &&
    condition3)
{
    ...
}
```

File: opendmarc/opendmarc.c [hunk 12]
```diff
@@ -3162,15 +3213,33 @@ mlfi_eom(SMFICTX *ctx)
...
+		&& conf->conf_dolog)
+	{
+		syslog(LOG_NOTICE,
+				"%s: ARC pass, policy fail > continuing DMARC eval",
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Alignment.

File: opendmarc/opendmarc.c [hunk 13]
```diff
@@ -3162,15 +3213,33 @@ mlfi_eom(SMFICTX *ctx)
...
+	}
+
+	if (result == DMARC_RESULT_REJECT && dfc->mctx_arcpolicypass == DMARC_ARC_POLICY_RESULT_PASS)
 	{
 		ret = SMFIS_ACCEPT;
 		result = DMARC_RESULT_ACCEPT;
 		if (conf->conf_dolog)
 		{
 			syslog(LOG_NOTICE,
-			       "%s: overriding DMARC fail due to ARC pass",
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Alignment.

File: opendmarc/opendmarc.c (hunk 8 above)
Comment (markeissler) [reply opendmarc/opendmarc.c]: Revisit. This is a logical operation. It should be, it should be a serial op.

Comment (markeissler) [reply opendmarc/opendmarc.c]: L2528 Is this logic correct? We need to pass on cv=none if we are the last recipient.

**Comment by markeissler** [comment]
Closing for collapsed stack:

* [PR#15 Add support for selectors, domain whitelist, arc.chain #15](https://github.com/trusteddomainproject/OpenDMARC/pull/15)

**Pull request closed without merging**